### ISSUE TITLE
Thai language weird rendering

### ISSUE THREAD

**Issue by pawawat** [mintty/mintty#1285]
I've noticed that the Thai language and its accent marks have a weird rendering. You can regenerate with the command:

`curl https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-demo.txt`

![image](https://github.com/user-attachments/assets/105f7029-0cdb-4e70-8ce9-50f982960545)

![image](https://github.com/user-attachments/assets/53890e8f-0292-4f8c-94f0-e70e7851929a)

The dotted symbol is overlapped

BTW, I use MonoLisa Font

**Comment by BrianInglis** [comment]
What did you expect to see, what do you actually see, and why is that significant in your view? 
Is the circle above the dotted circle meant to be combined and rendered above the following character? 
What about the dotted circles under other characters? 

Those with native understanding of a language and rendering can help us understand and improve its use by software. 
[You might also consider providing Thai translations of the text strings used by mintty based on those in `/usr/share/mintty/lang/messages.pot`, so you can use the package in your own language, and the developer can provide more details how you can help if you wish.]

WJFFM and aligns properly in mintty and in my browser with DejaVu Sans Mono and various backup/fallback/substitution fonts including Unifoundry Unifont and Unicode Last Resort (available as Cygwin packages) [magnified five `ctrl+`]:
![mintty Thai](https://github.com/user-attachments/assets/17818d74-a262-4cad-91b9-2431a72b95c6)

**Comment by mintty** [comment]
Taking just the last line and opening in notepad, I guess this is how it should look:
![grafik](https://github.com/user-attachments/assets/16032473-8ff9-4e5e-bd8c-7f962ccf6397)
Please confirm.

Mintty has a workaround for display problems of combining characters which does not seem to be proper for all scripts. I'll check.

**Comment by mintty** [comment]
There are 3 different issues; if I disable the workaround mentioned and let Windows render all combining characters, it looks already much better (first Thai line in screenshot) but still with a dotted circle glitch in the middle.
This glitch is around character U+0E33 which Unicode lists as compatibility character. If I decompose it, the result looks much better, with only the two accents above the U+0E04 KHO KHWAI exchanged, following the order of the two combining accents which are now separate. If I exchange them, of course the order is now as expected (last line).
Question: Is the order of accents maybe exchangeable in the Thai script? Or can you point out some rules by which such accents could be reordered automatically?
![grafik](https://github.com/user-attachments/assets/e19ac85f-beed-4013-892c-cede3770a102)

**Comment by pawawat** [comment]
> Taking just the last line and opening in notepad, I guess this is how it should look: ![grafik](https://private-user-images.githubusercontent.com/12740416/362584065-16032473-8ff9-4e5e-bd8c-7f962ccf6397.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU4MDcyNTUsIm5iZiI6MTcyNTgwNjk1NSwicGF0aCI6Ii8xMjc0MDQxNi8zNjI1ODQwNjUtMTYwMzI0NzMtOGZmOS00ZTVlLWJkOGMtN2Y5NjJjY2Y2Mzk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA4VDE0NDkxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkNTBkYmVjNWRiMzE1ZDg3YTY5NDgzNmZiODRkYTk5MjNiNjZmZDEzN2QwYTk4YjRiYTI4Y2RlNzE5NDY3ZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5yoMKSiiZ2f6VXEm6v7-iGBRl2zDij-ejHsC205LqkU) Please confirm.
> 
> Mintty has a workaround for display problems of combining characters which does not seem to be proper for all scripts. I'll check.

yes it should look like that


and as you can see  
![image](https://github.com/user-attachments/assets/9bf96c91-b7fc-493e-aca3-aae98f12ff4c)
The [Karan ](https://en.wikipedia.org/wiki/Thai_script#Diacritics)  is rendered twice, because ligature or something. I'm not sure

**Comment by pawawat** [comment]
![image](https://github.com/user-attachments/assets/26170ea3-1611-4a5a-b59e-d88313e198de)
the last one is correct, 

The last one is correct, but I'm not sure if the order should be ignored or auto-corrected.

**Comment by pawawat** [comment]
> There are 3 different issues; if I disable the workaround mentioned and let Windows render all combining characters, it looks already much better (first Thai line in screenshot) but still with a dotted circle glitch in the middle. This glitch is around character U+0E33 which Unicode lists as compatibility character. If I decompose it, the result looks much better, with only the two accents above the U+0E04 KHO KHWAI exchanged, following the order of the two combining accents which are now separate. If I exchange them, of course the order is now as expected (last line). 

What is the option to let Windows render all combining characters

> Question: Is the order of accents maybe exchangeable in the Thai script? Or can you point out some rules by which such accents could be reordered automatically? 

![image](https://github.com/user-attachments/assets/1c01500c-945b-4cad-95c7-ccfe801bab97)

It should be typed in this order.

**Comment by BrianInglis** [comment]
Does Windows Uniscribe have any preference for certain [Unicode Normalization](https://www.unicode.org/charts/normalization/) forms?
Could you try the alternatives easily e.g in Python using [unicodedata.normalize](https://docs.python.org/3.9/library/unicodedata.html#unicodedata.normalize)?
Then try the alternatives in `mintty` to validate it works the same? 
Some SO articles suggest NFKC may produce better results with some scripts.

**Comment by mintty** [comment]
I had suggested some kind of normalization would help but it turns out that is not necessary.
I have a working proof-of-concept patch meanwhile.
More tomorrow...

**Comment by mintty** [comment]
I had observed that Windows does not render certain combining character combinations and made them be rendered separately for mintty 2.4.0. I have no idea why this is the case, maybe Windows accepts only combined combinations that it thinks to know. Anyway, obviously the separate display strategy does not allow certain scripts to be rendered properly, especially where combined characters form merged glyphs, not just overlays.
The separate display is still appropriate, as I tested, for Latin scripts, and also for CJK.
So one question to be investigated is which scripts have combined character appearances that need to be render in onoe chunk via Windows, so mintty can make a clear distincition for one or the other approach.
Changing this experimentally already renders most of the text as it should - with one notable exception:
The character U+0E33 still produced garbage. It turns out this is due to a very special and weird property of only 4 Unicode characters, which have a compatibility decomposition that begins with a combining character:
```
0E33;THAI CHARACTER SARA AM;Lo;0;L;<compat> 0E4D 0E32;;;;N;THAI VOWEL SIGN SARA AM;;;;
0EB3;LAO VOWEL SIGN AM;Lo;0;L;<compat> 0ECD 0EB2;;;;N;;;;;
0F77;TIBETAN VOWEL SIGN VOCALIC RR;Mn;0;NSM;<compat> 0FB2 0F81;;;;N;;;;;
0F79;TIBETAN VOWEL SIGN VOCALIC LL;Mn;0;NSM;<compat> 0FB3 0F81;;;;N;;;;;
```
These need addtiaionl special handling.
Also this suggests that at least for Thai, Lao, and Tibetan scripts, combined characters should be handled by Windows, not rendered separately, but probably more, maybe all South East Asian scripts...
Any more insights or hints about this weird issue are appreciated.

**Comment by BrianInglis** [comment]
[MS Typography Thai Script Development](https://learn.microsoft.com/en-us/typography/script-development/thai) describes how Uniscribe should handle Thai script, shows font tables and options, and has a number of Feedback option links, from which you may get a response if you have a MS account or provide an email address

**Comment by mintty** [comment]
Please check out rendering for South East Asian scripts, now in the repository.
I'm applying the revised approach also for Indic scripts right now.
![grafik](https://github.com/user-attachments/assets/b1d24b15-b72f-49ea-b2e3-590490ccd50b)

**Comment by mintty** [comment]
Released 3.7.5.
